### PR TITLE
added unit tests for athena-postgresql.

### PR DESCRIPTION
**Pull request by ritiktrianz**
*Issue #, if available:*

*Description of changes:*
Added unit tests for athena-postgresql connector.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: athena-postgresql/src/test/java/com/amazonaws/athena/connectors/postgresql/PostGreSqlMetadataHandlerTest.java [hunk 1]
```diff
@@ -210,6 +245,89 @@ public void doGetTableLayout()
         Mockito.verify(preparedStatement, Mockito.times(1)).setString(2, tableName.getTableName());
     }
 
+    @Test
+    public void testListTablesWithValidConnection() {
```
Comment (AbdulR3hman) [inline athena-postgresql/src/test/java/com/amazonaws/athena/connectors/postgresql/PostGreSqlMetadataHandlerTest.java]: throw exception here; try/catch is not needed. let the exception bubble.

File: athena-postgresql/src/test/java/com/amazonaws/athena/connectors/postgresql/PostGreSqlMetadataHandlerTest.java [hunk 2]
```diff
@@ -210,6 +245,89 @@ public void doGetTableLayout()
...
+            Mockito.verify(matViewStatement).executeQuery();
+        } catch (Exception e) {
+            fail("Unexpected exception occurred: " + e.getMessage());
+        }
+    }
+
+
+    @Test
+    public void testListTablesWithNoMaterializedViews() {
+        try {
```
Comment (AbdulR3hman) [inline athena-postgresql/src/test/java/com/amazonaws/athena/connectors/postgresql/PostGreSqlMetadataHandlerTest.java]: same here; please let the exception bubble up.

File: athena-postgresql/src/test/java/com/amazonaws/athena/connectors/postgresql/PostGreSqlMetadataHandlerTest.java [hunk 3]
```diff
@@ -210,6 +245,89 @@ public void doGetTableLayout()
...
+    public void testListTablesWithValidConnection() {
+        try {
+            String databaseName = "testSchema";
+            Connection jdbcConnection = Mockito.mock(Connection.class);
+            PreparedStatement tablesStatement = Mockito.mock(PreparedStatement.class);
+            PreparedStatement matViewStatement = Mockito.mock(PreparedStatement.class);
+            ResultSet tablesResultSet = Mockito.mock(ResultSet.class);
+            ResultSet matViewResultSet = Mockito.mock(ResultSet.class);
+
+            when(jdbcConnection.prepareStatement(anyString())).thenReturn(tablesStatement, matViewStatement);
```
Comment (AbdulR3hman) [inline athena-postgresql/src/test/java/com/amazonaws/athena/connectors/postgresql/PostGreSqlMetadataHandlerTest.java]: this could lead to vagueness when it failed; giving that we are saying any string in the preprared statement; when that's not entirely correct; See below; where we know exactly what the string to list tables is. 

https://github.com/awslabs/aws-athena-query-federation/blob/de87534712adb9befb4cd4e3b65ec428c6403cb4/athena-jdbc/src/main/java/com/amazonaws/athena/connectors/jdbc/manager/JDBCUtil.java#L155

File: athena-postgresql/src/test/java/com/amazonaws/athena/connectors/postgresql/PostGreSqlMetadataHandlerTest.java [hunk 4]
```diff
@@ -210,6 +245,89 @@ public void doGetTableLayout()
...
+
+            List<TableName> result = postGreSqlMetadataHandler.listTables(jdbcConnection, databaseName);
+
+            List<TableName> expected = ImmutableList.of(
+                    new TableName(databaseName, "table1"),
+                    new TableName(databaseName, "matview1")
+            );
+
+            assertEquals("Expected tables and materialized views", expected, result);
+            Mockito.verify(tablesStatement).executeQuery();
```
Comment (AbdulR3hman) [inline athena-postgresql/src/test/java/com/amazonaws/athena/connectors/postgresql/PostGreSqlMetadataHandlerTest.java]: I believe we should verify ahead of testing for equality. given that would be the product of a correct execution order that is confirmed by the verify.

File: athena-postgresql/src/test/java/com/amazonaws/athena/connectors/postgresql/PostGreSqlMetadataHandlerTest.java [hunk 5]
```diff
@@ -503,6 +620,166 @@ public void doGetTableMaterializedView()
...
+
+        List<OptimizationSubType> hints = capabilities.get(DATA_SOURCE_HINTS);
+        assertNotNull("Expected " + DATA_SOURCE_HINTS + " capability to be present", hints);
+        assertEquals(HINTS_SIZE, hints.size());
+        Assert.assertTrue(hints.stream().anyMatch(subType -> subType.getSubType().equals(NON_DEFAULT_COLLATE)));
+    }
+
+    @Test
+    public void testGetCharColumnsWithCharacterColumns() {
+        try {
```
Comment (AbdulR3hman) [inline athena-postgresql/src/test/java/com/amazonaws/athena/connectors/postgresql/PostGreSqlMetadataHandlerTest.java]: same here;

**Review by chngpe — APPROVED** [review]

File: athena-postgresql/src/test/java/com/amazonaws/athena/connectors/postgresql/PostGreSqlMetadataHandlerTest.java (hunk 3 above)
Comment (Jithendar12) [reply athena-postgresql/src/test/java/com/amazonaws/athena/connectors/postgresql/PostGreSqlMetadataHandlerTest.java]: Added constants TABLES_SQL and MATERIALIZED_VIEWS_SQL at class level with exact SQL queries used in the implementation.

File: athena-postgresql/src/test/java/com/amazonaws/athena/connectors/postgresql/PostGreSqlMetadataHandlerTest.java (hunk 4 above)
Comment (Jithendar12) [reply athena-postgresql/src/test/java/com/amazonaws/athena/connectors/postgresql/PostGreSqlMetadataHandlerTest.java]: Moved Mockito.verify() calls before assertions.

File: athena-postgresql/src/test/java/com/amazonaws/athena/connectors/postgresql/PostGreSqlMetadataHandlerTest.java (hunk 5 above)
Comment (Jithendar12) [reply athena-postgresql/src/test/java/com/amazonaws/athena/connectors/postgresql/PostGreSqlMetadataHandlerTest.java]: Removed try/catch blocks and added throws SQLException.

File: athena-postgresql/src/test/java/com/amazonaws/athena/connectors/postgresql/PostGreSqlMetadataHandlerTest.java (hunk 2 above)
Comment (Jithendar12) [reply athena-postgresql/src/test/java/com/amazonaws/athena/connectors/postgresql/PostGreSqlMetadataHandlerTest.java]: Removed try/catch blocks and added throws SQLException.

File: athena-postgresql/src/test/java/com/amazonaws/athena/connectors/postgresql/PostGreSqlMetadataHandlerTest.java (hunk 1 above)
Comment (Jithendar12) [reply athena-postgresql/src/test/java/com/amazonaws/athena/connectors/postgresql/PostGreSqlMetadataHandlerTest.java]: Thank you for reviewing @AbdulR3hman.
Removed try/catch blocks and added throws SQLException.

**Comment by burhan94** [comment]
Can we consider using GenAI here to add all of the missing cases and add even more robust UT's? We should be able to get 30-50 new test cases here.

**Comment by ritiktrianz** [comment]
Hi @burhan94 ,Thank you for the feedback.  We already tried GenAI using Claude Sonnet models, for unit test generation. It generated large volume of test cases covering multiple scenarios, such as edge cases, negative paths, and broader input variations. We felt that not all of these test cases may be relevant and may create unnecessary overhead. So, we removed a few tests and kept only the relevant ones.
 
However, now We have added the generated tests after reviewing them manually.

**Comment by burhan94** [comment]
Can you add some UT's to cover the postgres SQL generation from the constraints object? Basically create a bunch of different scenarios that test every predicate type for postgres that we support today. I'd like to see all of the below:
1. Complex expressions
2. Top N
3. Limit
4. order by

**Comment by ritiktrianz** [comment]
Hi @burhan94 , thank you for the review comment. We’ve added unit tests to cover all supported predicate types for Postgres.

File: athena-postgresql/src/test/java/com/amazonaws/athena/connectors/postgresql/PostGreSqlMetadataHandlerTest.java [hunk 6]
```diff
@@ -205,9 +266,89 @@ public void doGetTableLayout()
...
+        when(matViewStatement.executeQuery()).thenReturn(emptyResultSet);
+
+        List<TableName> result = postGreSqlMetadataHandler.listTables(mockConn, TEST_SCHEMA);
+
+        List<TableName> expected = ImmutableList.of(new TableName(TEST_SCHEMA, "table1"));
+        assertEquals("Expected only tables", expected, result);
+    }
+
+    @Test
+    public void testListTablesWithDatabaseConnectionError() {
```
Comment (ejeffrli) [inline athena-postgresql/src/test/java/com/amazonaws/athena/connectors/postgresql/PostGreSqlMetadataHandlerTest.java]: can we not use try-catch here for this test?

Comment (ritiktrianz) [reply athena-postgresql/src/test/java/com/amazonaws/athena/connectors/postgresql/PostGreSqlMetadataHandlerTest.java]: Hi @ejeffrli , As suggested, we've removed the try-catch block from the test and replaced it with assertThrows for cleaner exception handling. Please review and let us know if you have any further feedback.

File: athena-postgresql/src/test/java/com/amazonaws/athena/connectors/postgresql/PostGreSqlEnvironmentPropertiesTest.java [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+        connectionProperties.put(SECRET_NAME, "postgres-secret");
+        connectionProperties.put(PORT, DEFAULT_PORT);
+        postGreSqlEnvironmentProperties = new PostGreSqlEnvironmentProperties();
+    }
+
+    @Test
+    public void testBasicConnectionString() {
+        Map<String, String> postGreSqlConnectionProperties = postGreSqlEnvironmentProperties.connectionPropertiesToEnvironment(connectionProperties);
+
+        String expectedConnectionString = JDBC_URL_PREFIX + DEFAULT_HOST + ":" + DEFAULT_PORT + "/" + DEFAULT_DATABASE + "?" + SECRET_PLACEHOLDER;
```
Comment (ejeffrli) [inline athena-postgresql/src/test/java/com/amazonaws/athena/connectors/postgresql/PostGreSqlEnvironmentPropertiesTest.java]: nit: For expected string assertions, can we write out the whole string explicitly? for example, it would be "postgres://jdbc:postgresql://..." This way developers can quickly verify differences.

Comment (ejeffrli) [reply athena-postgresql/src/test/java/com/amazonaws/athena/connectors/postgresql/PostGreSqlEnvironmentPropertiesTest.java]: This is more readable
```
 assertEquals(
        "postgres://jdbc:postgresql://postgres-endpoint:5000/testdb?${postgres-secret}",
        postGreSqlConnectionProperties.get(DEFAULT)
    );
    ```

Comment (ritiktrianz) [reply athena-postgresql/src/test/java/com/amazonaws/athena/connectors/postgresql/PostGreSqlEnvironmentPropertiesTest.java]: > nit: For expected string assertions, can we write out the whole string explicitly? for example, it would be "postgres://jdbc:postgresql://..." This way developers can quickly verify differences.

Hi @ejeffrli , thank you for your suggestion. We’ve updated the assertions to improve readability. Please review and let us know if you have any further feedback.

File: athena-postgresql/src/test/java/com/amazonaws/athena/connectors/postgresql/PostgreSqlFederationExpressionParserTest.java [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+
+    @Test
+    public void testWriteArrayConstructorClauseWithSpecialCharacters() {
+        List<String> arguments = Arrays.asList("'Hello, World!'", "'It''s working'", "'Semi;Colon'");
+        String result = parser.writeArrayConstructorClause(new ArrowType.Utf8(), arguments);
+        assertEquals("'Hello, World!', 'It''s working', 'Semi;Colon'", result);
+    }
+
+    @Test
+    public void testWriteArrayConstructorClauseWithNull() {
```
Comment (ejeffrli) [inline athena-postgresql/src/test/java/com/amazonaws/athena/connectors/postgresql/PostgreSqlFederationExpressionParserTest.java]: is this actually testing null? Seems like null is the same as any string here

Comment (ritiktrianz) [reply athena-postgresql/src/test/java/com/amazonaws/athena/connectors/postgresql/PostgreSqlFederationExpressionParserTest.java]: > is this actually testing null? Seems like null is the same as any string here

Hi @ejeffrli , thank you. We’ve replaced the null string test with an actual null argument test. Kindly review.

Comment (ejeffrli) [reply athena-postgresql/src/test/java/com/amazonaws/athena/connectors/postgresql/PostgreSqlFederationExpressionParserTest.java]: I'm not sure if this is the right change. I believe arguments should be SQL examples. In this case, would it be a SQL null? That means it would be a string like "null"

Let me know if there was another intent with this test.

Comment (ritiktrianz) [reply athena-postgresql/src/test/java/com/amazonaws/athena/connectors/postgresql/PostgreSqlFederationExpressionParserTest.java]: > I'm not sure if this is the right change. I believe arguments should be SQL examples. In this case, would it be a SQL null? That means it would be a string like "null"
> 
> Let me know if there was another intent with this test.

Hi @ejeffrli , thanks for the review. We verified further and confirmed that there is a null check earlier in the flow, so this method would never receive a Java null. The original intent of this test was to cover a negative scenario where "null" is treated as a string value. Since similar coverage already exists, we have removed the SQL "null" test as well.
Please review and let us know if you have any feedback.

Comment (ejeffrli) [reply athena-postgresql/src/test/java/com/amazonaws/athena/connectors/postgresql/PostgreSqlFederationExpressionParserTest.java]: which test is the similar coverage?

Comment (VenkatasivareddyTR) [reply athena-postgresql/src/test/java/com/amazonaws/athena/connectors/postgresql/PostgreSqlFederationExpressionParserTest.java]: Hi @ejeffrli, these two tests provide the coverage we mentioned for String types, as "null" value also treated as String type only. Here are the reference links for each:
https://github.com/awslabs/aws-athena-query-federation/pull/2822/files#diff-fc042a21655b9e3353e4024d4c649d3426eefc86952b5a8e7ea19434844485cfR43

https://github.com/awslabs/aws-athena-query-federation/pull/2822/files#diff-fc042a21655b9e3353e4024d4c649d3426eefc86952b5a8e7ea19434844485cfR64
